nx/lib/nx/defn/compilation_diff.ex

```diff
@@ -0,0 +1,96 @@
+defmodule Nx.Defn.CompilationDiff do
+  defstruct [:left, :right, :compatible, :nesting]
+
+  defp is_valid_container?(t) do
+    impl = Nx.Container.impl_for(t)
```

Comment:
You are getting the container several times. Get it once at the top of `build` for both `left` and `right` and then pass it around. :)